src/prog_algs/state_estimators/particle_filter.py

```diff
@@ -52,45 +55,71 @@ def __init__(self, model, x0, measurement_eqn = None, **kwargs):
         # State-estimator specific logic
         if isinstance(self.parameters['n'], Number):
             self.parameters['n'] = {key : self.parameters['n'] for key in self.__measure(x0).keys()}
+        # todo(CT): Check fields on n
+        
+        # Build array inplace
+        x = array(list(x0.values()))
 
-        if isinstance(self.parameters['x0_uncertainty'], Number):
-            # Build array inplace
-            x = array(list(x0.values()))
+        if isinstance(self.parameters['x0_uncertainty'], dict):
+            sd = array(list(self.parameters['x0_uncertainty'].values()))
+        elif isinstance(self.parameters['x0_uncertainty'], Number):
             sd = array([self.parameters['x0_uncertainty']] * len(x0))
-            samples = array([random.normal(x, sd) for i in range(self.parameters['num_particles'])])
-            self.particles = array([{key: value for (key, value) in zip(x0.keys(), x)} for x in samples])
         else:
             raise ProgAlgTypeError
+
+        samples = [random.normal(
+            x[i], sd[i], self.parameters['num_particles']) for i in range(len(x))]
+        self.particles = dict(zip(x0.keys(), samples))
     
     def __str__(self):
         return "{} State Estimator".format(self.__class__)
         
     def estimate(self, t, u, z):
+        assert t > self.t, "New time must be greater than previous"
         dt = t - self.t
         self.t = t
-        weights = empty(len(self.particles))
-        
+
         # Optimization
         particles = self.particles
         next_state = self.model.next_state
         apply_process_noise = self.model.apply_process_noise
         output = self.__measure
+        apply_measurement_noise = self.model.apply_measurement_noise
         noise_params = self.parameters['n']
 
-        # Propogate and calculate weights
-        for i, item in enumerate(particles):
-            particles[i] = next_state(item, u, dt) 
-            particles[i] = apply_process_noise(particles[i])
-            zPredicted = output(particles[i])
-            weights[i] = sum([norm(zPredicted[key], noise_params[key]).pdf(z[key]) for key in zPredicted.keys()])
-        
+        # Propagate particles state
+        self.particles = apply_process_noise(next_state(particles, u, dt))
+
+        # Get particle measurements
+        zPredicted = apply_measurement_noise(output(self.particles))
+
+        # Calculate pdf values
+        pdfs = array([norm(zPredicted[key], noise_params[key]).logpdf(z[key])
+                      for key in zPredicted.keys()])
+
+        # Calculate log weights
+        log_weights = pdfs.sum(0)
+
+        # Scale
```

Comment:
```suggestion
        # Scale
        # We subtract the max log weights for numerical stability. 
        # Sometimes log weights can be a large negative value
        # when you exponentiate that value the computer will round the result to 0 for most of the weights (sometimes all of them) 
        # this causes problems when trying to sample from the particles. 
        # We shift them up by the max log weight (essentially making the max log weight 0) to help avoid that problem. 
        # When we normalize the weights by dividing by the sum of all the weights, that constant cancels out.
```

Added your description as a comment for others users